### DISCUSSION TITLE
"Profiling" feature mentioned on the PF SOC HSS GitHub page - any more details?

### DISCUSSION THREAD

**Discussion by haroldnelson-2021** [polarfire-soc/polarfire-soc-documentation#303] [HSS and U-Boot]
Are there more detailed instructions and-or a user example regarding the "Profiling" feature mentioned on the PF SOC HSS GitHub page?

**Comment by griffini** [comment]
The profiling mentioned in the README is specifically concerned with internal profiling of the HSS. There is a Kconfig option (`CONFIG_DEBUG_PROFILING_MAX_NUM_FUNCTIONS`) which is used to control how many internal functions can be traced.

To enable profiling on a file-by-file basis, you can add custom rules to the appropriate Makefile to add `-finstrument-functions` to the command line for that file. Alternatively, you can add `__attribute__((instrument_function))` before a function to cause it to be instrumented.

Perhaps more useful are the profile counters, in `modules/debug/hss_perfctr.h`. These are enabled using the `CONFIG_DEBUG_PROFILING_SUPPORT` Kconfig option. These allow timers to be added in the code to wallclock time various operations.  If you type `debug perfctr` into the HSS, you'll see (depending on your boot configuration) something like the following output:

````
>> debug perfctr
[343873.275275]     4965 ms ( 4964969 ticks) - DDR Init
[343873.281351]        0 ms (      60 ticks) - MemTest(DDR32)
[343873.287948]        0 ms (      60 ticks) - MemTest(DDR64)
[343873.294545]      413 ms (  412959 ticks) - MMC Init
[343873.300621]      423 ms (  422829 ticks) - Boot Image Init
[343873.307305]      360 ms (  360194 ticks) - MMC Init
[343873.313381]        9 ms (    9256 ticks) - Boot Image MMC Copy
[343873.320411]      344 ms (  343730 ticks) - boot_service(u54_1)
[343873.327442]      183 ms (  182818 ticks) - boot_service(u54_2)
[343873.334473]      177 ms (  176569 ticks) - boot_service(u54_3)
[343873.341504]      170 ms (  170321 ticks) - boot_service(u54_4)
````

All of these counters are timing sequential events, except for the 4 `boot_service` counters which effectively count time in parallel.  So the boot time to payload handoff is the sum of `DDR Init` through `Boot Image MMC Copy` *plus* the largest of the 4 `boot_service` counters.

**Comment by rockoguido** [comment]
Hello, as a follow-on to this thread this is the exact scenario I am trying to get to work, for the HSS targeting an Icicle Kit.  I would like to get profiling working when building up the latest version of HSS,  2203.02.1.    I can build the image without an issue but then when enabling the option referenced above (CONFIG_DEBUG_PROFILING_SUPPORT) I get 3 undefines at link time..  g5soc_init_mutex, g5soc_take_mutex and g5soc_release_mutex.    The errors make sense when looking at the code since they are referenced but not defined.   It seems an extra library or source code needs to be brought in somehow..   
As an aside when using Kconfig (make menuconfig against the .config) under Debug options the profiling options does not even appear, I can change it manually by editing .config directly but this makes me uncomfortable.   
Hoping it's just pilot error on my part..
Thx!
Dave

**Comment by griffini** [comment]
Hi @rockoguido  - it was definitely not pilot error on your part. We've just pushed [a fix](https://github.com/polarfire-soc/hart-software-services/commit/7610384877c2e9d1aa58a0c5e0188f120d70a146) to enable profiling in the HSS.

Edit the .config manually and it should build okay for you. For documentation including how to use this, see https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/hss-and-u-boot/profiling.md
